### ISSUE TITLE
Autofix inadvertently changing code outside the match

### ISSUE THREAD

**Issue by msorens** [semgrep/semgrep#1721]
**Describe the bug**
Autofix properly updated the matched pattern with the replacement text--but it also changed the code outside the match and it should not have. I first found this by running semgrep on the command line with a YAML file; then reproduced precisely the same behavior at semgrep.dev.

**To Reproduce**
In the playground (https://semgrep.dev/s/7g4k/), run the sample and apply the autofix. It correctly changes the match but also deletes the immediately adjacent semicolon.

**Expected behavior**
Surrounding code should be 100% preserved.

**Screenshots**
![image](https://user-images.githubusercontent.com/6817500/93963640-92a8dc00-fd12-11ea-9b3a-f2cf68cfd591.png)

**Environment**
v 0.24.0

**Comment by chmccreery** [comment]
Found the same bug in python: https://semgrep.dev/s/LrkX/

**Comment by brendongo** [comment]
Looks like we might have some disparity on how UI and CLI treat end column (inclusive vs exclusive). I think the webUI is treating it as exclusive, but the autofix code treats it as inclusive. Will have to double check with semgrep-core on what actual intention is.

Evidence:
https://semgrep.dev/s/RrNv/?stepByStep=1 shows finding range as 0-5
WebUI highlights 0-4 but autofix replaces characters 0-5.

**Comment by nbrahms** [comment]
cc. @aryx

**Comment by aryx** [comment]
Here is the code in semgrep-core/reporting/JSON_report.ml
```
let json_range min max =
  (* pfff (and Emacs) have the first column at index 0, but not r2c *)
  let adjust_column x = x + 1 in

  let min = PI.token_location_of_info min in
  let max = PI.token_location_of_info max in
  let len_max = String.length max.PI.str in
  J.Object [
    "line", J.Int min.PI.line;
    "col", J.Int (adjust_column min.PI.column);
    "offset", J.Int min.PI.charpos;
  ],
  J.Object [
    "line", J.Int max.PI.line;
    "col", J.Int (adjust_column (max.PI.column + len_max));
    "offset", J.Int (max.PI.charpos + len_max);
  ]

```

**Comment by aryx** [comment]
So it's adding the length of the matched thing to the starting offset, so I guess this is exclusive in brendon's terminology.
On brendon's example, here is what semgrep-core report:
```
+ /home/pad/semgrep/_build/default/bin/Main.exe -lang python -rules_file test.yaml test.py
{"matches":[{"check_id":"boto3-internal-network","path":"test.py","start":{"line":1,"col":1,"offset":0},"end":{"line":1,"col":6,"offset":5},"extra":{"message":"Boto3 connection to internal network\n","metavars":{}}}],"errors":[],"stats":{"okfiles":1,"errorfiles":0}}
```

**Comment by aryx** [comment]
So boto3 in the code starts at offset 0, and its length is 5, so the end offset reported is 5, but indeed the character at offset 5 is not part of the matched string (hence the use of 0-4 in highlighting).

**Comment by dlukeomalley** [comment]
@msorens This should be resolved by #1859 once it merges. Thanks @brendongo for leading the charge on the fix.

**Comment by DrewDennison** [comment]
https://r2c-community.slack.com/archives/CK86BJ5DW/p1603471048306900
